libraries/image/src/utils/webcam.ts

```diff
@@ -15,49 +15,118 @@
  * =============================================================================
  */
 
+import autobind from 'autobind-decorator';
+
 const defaultVideoOptions: MediaTrackConstraints = {
     facingMode: 'user',
     frameRate: 30,
     aspectRatio: 1
 };
 
-/**
- * utility to get a webcam feed in a video element
- * @returns Promise<HTMLVideoElement>
- */
-export function getWebcam(
-    width = 400,
-    height = 400,
-    facingMode = 'user',
-    flipped = true,
-    video: HTMLVideoElement= document.createElement('video'),
-    options: MediaTrackConstraints= defaultVideoOptions ) {
-    if (!window.navigator.mediaDevices || !window.navigator.mediaDevices.getUserMedia) {
-        return Promise.reject('Your browser does not support WebRTC. Please try another one.');
-    }
-
-    defaultVideoOptions.width = width;
-    defaultVideoOptions.height = height;
-
-    if (facingMode.toLowerCase() === 'back') {
-        defaultVideoOptions.facingMode = 'environment';
-    }
-
-    // should be flipped since we trained with flipped webcam
-    // but user can still change
-    if (flipped) {
-        const rotate = (value: string) =>
-            `rotateY(${value}); -webkit-transform:rotateY(${value}); -moz-transform:rotateY(${value})`;
-        video.setAttribute('style', `transform: ${rotate('180deg')}`);
-    }
-
-    return window.navigator.mediaDevices.getUserMedia({ video: options })
-        .then((mediaStream) => {
-            video.srcObject = mediaStream;
-            video.width = width;
-            video.height = height;
-            return video;
-        }, () => {
-            return Promise.reject('Could not open your camera. You may have denied access.');
+const fillConstraints = (options: Partial<MediaTrackConstraints>) => {
+    options.facingMode = options.facingMode || defaultVideoOptions.facingMode;
+    options.frameRate = options.frameRate || defaultVideoOptions.frameRate;
+    options.aspectRatio = options.aspectRatio || defaultVideoOptions.aspectRatio;
+    return options as MediaTrackConstraints;
+};
+
+export class Webcam  {
+    public flip: boolean;
+    public width: number;
+    public height: number;
+    public webcam: HTMLVideoElement;
+    public canvas: HTMLCanvasElement;
+    public mediaConstraints: MediaTrackConstraints;
+
+    constructor(width = 400, height = 400, flip = false, options: MediaTrackConstraints = {}) {
+        this.width = width;
+        this.height = height;
+        this.flip = flip;
+        this.mediaConstraints = options;
```

Comment:
this probably shouldnt come in the constructor. it is held as a member variable but only used in getWebcam and is also modified on 54 and 55. perhaps this is better as options for `getWebcam` than constructor